components/director/internal/open_resource_discovery/sanitizer.go

```diff
@@ -0,0 +1,455 @@
+package ord
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/kyma-incubator/compass/components/director/internal/model"
+	"github.com/pkg/errors"
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+)
+
+const unknownReferenceCode = "sap-ord-unknown-reference"
+const duplicateResourceCode = "sap-ord-duplicate-resource"
+
+// DocumentSanitizer represents the sanitizer of ORD documents
+type DocumentSanitizer struct {
+}
+
+// NewDocumentSanitizer returns new instance of a document sanitizer
+func NewDocumentSanitizer() DocumentSanitizer {
+	return DocumentSanitizer{}
+}
+
+// Sanitize performs all the merging and rewriting rules defined in ORD. This method should be invoked after Documents are validated with the Validate method.
+//   - Rewrite all relative URIs using the baseURL from the Described System Instance. If the Described System Instance baseURL is missing the provider baseURL (from the webhook) is used.
+//   - Package's partOfProducts, tags, countries, industry, lineOfBusiness, labels are inherited by the resources in the package.
+//   - Ensure to assign `defaultEntryPoint` if missing and there are available `entryPoints` to API's `PartOfConsumptionBundles`
+//   - If some resource(Package, API, Event or Data Product) doesn't have provided `policyLevel` and `customPolicyLevel`, these are inherited from the document
+func (v *DocumentSanitizer) Sanitize(docs []*Document, webhookBaseURL, webhookBaseProxyURL string) ([]*ValidationError, error) {
+	valErrors := make([]*ValidationError, 0)
+
+	var err error
+
+	// Use the ProxyURL for all relative link substitution except for the API's TargetURLs.
+	// They are externally consumable and we should not expose those URLs through the Proxy but rather from webhook's BaseURL
+	url := webhookBaseURL
+	if webhookBaseProxyURL != "" {
+		url = webhookBaseProxyURL
+	}
+
+	// Rewrite relative URIs
+	for _, doc := range docs {
+		for _, pkg := range doc.Packages {
+			if pkg.PackageLinks, err = rewriteRelativeURIsInJSON(pkg.PackageLinks, url, "url"); err != nil {
+				return valErrors, err
+			}
+			if pkg.Links, err = rewriteRelativeURIsInJSON(pkg.Links, url, "url"); err != nil {
+				return valErrors, err
+			}
+		}
+
+		for _, bndl := range doc.ConsumptionBundles {
+			if bndl.Links, err = rewriteRelativeURIsInJSON(bndl.Links, url, "url"); err != nil {
+				return valErrors, err
+			}
+			if bndl.CredentialExchangeStrategies, err = rewriteRelativeURIsInJSON(bndl.CredentialExchangeStrategies, url, "callbackUrl"); err != nil {
+				return valErrors, err
+			}
+		}
+
+		for _, api := range doc.APIResources {
+			for _, definition := range api.ResourceDefinitions {
+				if !isAbsoluteURL(definition.URL) {
+					definition.URL = url + definition.URL
```

Comment:
should we use a library for url concatenation here?

Reply:
Fixed with url.ResolveReference